Sources/ContentfulSyncSerializer/SyncJSONDownloader.swift

```diff
@@ -6,97 +6,137 @@
-
-        syncGroup = DispatchGroup()
     }
 
-    public func run(then completion: @escaping (Result<Bool>) -> Void) {
-
-        var clientConfiguration = ClientConfiguration()
-        clientConfiguration.dataDelegate = self
-        let client = Client(spaceId: spaceId,
-                            accessToken: accessToken,
-                            clientConfiguration: clientConfiguration)
+    public func run(then completion: @escaping (Contentful.Result<Bool>) -> Void) {
+        let client = Client(spaceId: spaceId,environmentId: environment,
+                            accessToken: accessToken)
 
-        print("Writing sync JSON files to directory \(outputDirectoryPath)")
-
-        client.sync { [unowned self] (result: Result<SyncSpace>) in
+        client.sync { [unowned self] (result: Contentful.Result<SyncSpace>) in
             guard let syncSpace = result.value, result.error == nil else {
                 completion(Result.error(result.error!))
                 return
             }
-            guard self.shouldDownloadMediaFiles && syncSpace.assets.count > 0 else {
-                completion(Result.success(true))
-                return
-            }
-            var imageSaveErrorCount = 0
-            for asset in syncSpace.assets {
-
-                self.syncGroup.enter()
-                client.fetchData(for: asset).then { data in
-                    do {
-                        try self.saveData(data, for: asset)
-                        self.syncGroup.leave()
-                    } catch {
-                        // TODO: Log error
-                        imageSaveErrorCount += 1
-                        self.syncGroup.leave()
+            self.fetchLocales(withClient: client) { [unowned self] result in
+                switch result {
+                case .success:
+                    self.fetchSync(withClient: client, syncSpace: syncSpace) { [unowned self] result in
+                        switch result {
+                        case .success:
+                            if self.shouldDownloadMediaFiles && syncSpace.assets.count > 0 {
+                                self.fetchAssets(withClient: client, syncSpace: syncSpace) { (result) in
+                                    if let error = result.error {
+                                        completion(.error(error))
+                                    }else {
+                                        completion(.success(true))
+                                    }
+                                }
+                            } else {
+                                completion(Result.success(true))
+                            }
+                        case let .error(error):
+                            completion(.error(error))
+                        }
                     }
+                case let .error(error):
+                    completion(.error(error))
+                }
+            }
+        }
+    }
```

Comment:
Is there anything we can do about the callback hell? 😅

Reply:
yeah, I can refactor it, to use some reactive way of handling